frame/contracts/src/migration/v9.rs

```diff
@@ -88,9 +107,35 @@ impl<T: Config> Migrate<T> for Migration<T> {
 			};
 			CodeStorage::<T>::insert(key, module);
 			self.last_key = Some(iter.last_raw_key().to_vec().try_into().unwrap());
-			(IsFinished::No, None)
+			(IsFinished::No, T::WeightInfo::v9_migration_step(len))
 		} else {
-			(IsFinished::Yes, None)
+			log::debug!(target: LOG_TARGET, "No more contracts code to migrate");
+			(IsFinished::Yes, T::WeightInfo::v9_migration_step(0))
 		}
 	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade_step() -> Result<Vec<u8>, &'static str> {
+		let sample: Vec<_> = old::CodeStorage::<T>::iter().take(10).collect();
+
+		log::debug!(target: LOG_TARGET, "Taking sample of {} contract codes", sample.len());
+		Ok(sample.encode())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade_step(state: Vec<u8>) -> Result<(), &'static str> {
+		let sample =
+			<Vec<(CodeHash<T>, old::PrefabWasmModule)> as Decode>::decode(&mut &state[..]).unwrap();
+
+		log::debug!(target: LOG_TARGET, "Validating sample of {} contract codes", sample.len());
+		for (code_hash, old) in sample {
```

Comment:
not sure what the good thing to test here, 

right now we run this right after the step, so it's unsafe to start exercising code from the pallet since there might be more steps that touch this before the end of the migration, so I ended up with these tests that just assert the result of the step function


An other option would be to un the post_upgrade_step after all the steps are performed, and then do something with the pallet code, but here again we might get into situation where some of our assumptions are false because following steps have touched some values

Reply:
I think this is fine for now. Anything more advanced comes with its own drawbacks as you pointed out. So let's just merge it as-is and iterate later.